core/src/main/java/org/opensearch/sql/planner/SerializablePlan.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.planner;
+
+import java.io.Externalizable;
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.io.Serializable;
+import org.apache.commons.lang3.NotImplementedException;
+import org.opensearch.sql.storage.StorageEngine;
+
+/**
+ * All instances of PhysicalPlan which needs to be serialized (in cursor feature) should
+ * override all given here methods.
+ * This class can't implement Externalizable, because deserialization of Externalizable objects
+ * works the following way:
+ * 1. A new object created with no-arg constructor (no PhysicalPlan has it)
+ * 2. Object loads data from the stream.
+ * {@link Externalizable} interface was split into two pars: serialization is kept with
+ * {@link #writeExternal}, but deserialization is provided by {@link PlanLoader}.
+ */
+public abstract class SerializablePlan {
+
+  // Copied from Externalizable
+  /**
+   *  Each plan which supports serialization should dump itself into the stream and go recursive.
+   *  It is good to create and dump a {@link PlanLoader} here as well. See usage samples.
+   *  <pre>{@code
+   *  out.writeSomething(data);
+   *  for (var plan : getChild()) {
+   *    plan.getPlanForSerialization().writeExternal(out);
+   *  }
+   *  }</pre>
+  */
+  public boolean writeExternal(ObjectOutput out) throws IOException {
+    throw new NotImplementedException();
+  }
+
+  /**
+   * Override to return child or delegated plan, so parent plan should skip this one
+   * for serialization, but it should try to serialize grandchild plan.
+   * Imagine plan structure like this
+   *    A         -> this
+   *    `- B      -> child
+   *      `- C    -> this
+   * In that case only plans A and C should be attempted to serialize.
+   * It is needed to skip a `ResourceMonitorPlan` instance only, actually.
+   * @return Next plan for serialization.
+   */
+  public SerializablePlan getPlanForSerialization() {
+    return this;
+  }
+
+  /**
+   * Each plan should serialize an instance of this function.
+   * The function deserializes and creates a new instance of that plan type.
+   * A loader of a plan X could be defined only in scope of X, because only X
+   * knows how to create a new X.
+   * Deserialization should match with serialization given in {@link #writeExternal}.
+   */
+  @FunctionalInterface
+  public interface PlanLoader extends Serializable {
+    SerializablePlan apply(ObjectInput in, StorageEngine engine)
+        throws IOException, ClassNotFoundException;
+  }
```

Comment:
I had two reasons of doing that:
1. Customized loader can have an extra argument (`engine`)
2. `readExternal` is called after creating a new object, so all serializable plan should have a no-arg constructor.

Reply:
1. This can be solved within serialization API. Please, see [`ObjectInputStream.resolveObject`](https://docs.oracle.com/javase/7/docs/api/java/io/ObjectInputStream.html#resolveObject(java.lang.Object)) or [`readResolve`](https://docs.oracle.com/javase/8/docs/technotes/guides/serialization/examples/symbol/Substitute.java).
2. As awkward as no-arg constructor are, everybody who knows Java serialization API expects them -- they are a requirement of the [serialization API](https://docs.oracle.com/javase/7/docs/platform/serialization/spec/serial-arch.html#7185). 
   - Not following the intended use of the API makes this code harder to maintain.
   - In addition, serializing the plan loader increases the size of the serialized plan and adds a possible vector for code injection attack.

Reply:
1. I'm not sure that I understand `resolveObject` purpose correctly, but I got it working. Please, see my fix in 490af96e8.
2. There is no problem to have a no-arg CTOR in `ProjectOperator`, `PaginateOperator` and so on, but `OpenSearchPagedIndexScan` requires some things to initialize which can't be deserialized. But I can use `resolveObject` the same way as I did to obtain them! Please, let me know whether I'm on the right track.